mobile/src/main/java/org/openhab/habdroid/core/connection/ConnectionFactory.java

```diff
@@ -0,0 +1,282 @@
+
+/**
+ * A factory class, which is the main entry point to get a Connection to a specific openHAB
+ * server. Use this factory class whenever you need to obtain a connection to load additional
+ * data from the openHAB server or another supported source (see the constants in {@link Connection}).
+ */
+final public class ConnectionFactory
+        extends BroadcastReceiver implements SharedPreferences.OnSharedPreferenceChangeListener {
+    private static final String TAG = ConnectionFactory.class.getSimpleName();
+    private static final List<Integer> localConnectionTypes = new ArrayList<>(
+            Arrays.asList(ConnectivityManager.TYPE_ETHERNET, ConnectivityManager.TYPE_WIFI,
+                    ConnectivityManager.TYPE_WIMAX));
+    private static final List<String> needInvalidateCachePreferenceKeys = Arrays.asList(Constants
+            .PREFERENCE_REMOTE_URL, Constants.PREFERENCE_LOCAL_USERNAME, Constants
+            .PREFERENCE_LOCAL_PASSWORD, Constants.PREFERENCE_REMOTE_PASSWORD, Constants
+            .PREFERENCE_REMOTE_USERNAME, Constants.PREFERENCE_LOCAL_URL, Constants.PREFERENCE_DEMOMODE);
+
+    private Context ctx;
+    private SharedPreferences settings;
+    private NetworkInfo activeNetworkInfo;
+
+    Map<Integer, Connection> cachedConnections = new HashMap<>();
+
+    private static class InstanceHolder {
+        public static final ConnectionFactory INSTANCE = new ConnectionFactory();
+    }
+
+    static ConnectionFactory getInstance() {
+        return InstanceHolder.INSTANCE;
+    }
+
+    ConnectionFactory() {
+        if (Build.VERSION.SDK_INT >= 21) {
+            localConnectionTypes.add(ConnectivityManager.TYPE_VPN);
+        }
+    }
+
+    private void setContext(Context ctx) {
+        this.ctx = ctx;
+    }
+
+    @Override
+    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
+        if (needInvalidateCachePreferenceKeys.contains(key)) {
+            cachedConnections.clear();
+        }
+    }
+
+    private void setSettings(SharedPreferences settings) {
+        if (this.settings != null) {
+            this.settings.unregisterOnSharedPreferenceChangeListener(this);
+        }
+        this.settings = settings;
+
+        this.settings.registerOnSharedPreferenceChangeListener(this);
+    }
+
+    public static Connection getConnection(int connectionType, Context ctx) {
+        return getConnection(connectionType, ctx,
+                PreferenceManager.getDefaultSharedPreferences(ctx));
+    }
+
+    public static Connection getConnection(int connectionType, Context ctx,
+                                           SharedPreferences settings) {
+        // FIXME: Once we're on API level 19, use Objects.requireNonNull();
+        if (ctx == null) {
+            throw new NullPointerException("Context ctx can not be null.");
+        }
+
+        ConnectionFactory factory = getInstance();
+        factory.setSettings(settings);
+        factory.setContext(ctx);
```

Comment:
This, together with the registration logic, is really ugly, as it e.g. might/will leak activity objects. I'd suggest to
- add an Application object
- initialize the factory instance in Application.onCreate, passing in the Application object (which equals the application context)
- fetch the SharedPreferences instance from the passed in application context and register the pref listener for it
- let the Application object register the broadcast receiver

This way you won't leak anything and can get rid of the unregistration logic. You'll also get rid of the need of passing a Context or SharedPreferences around.

Reply:
+1, thanks for the advice :)

Reply:
See my comment above, those setters can die ;-)

Reply:
Then I would need to make the ConnectionFactory constructor public and would require to make the singleton instance holder public, too, otherwise the Application object can not set the factory instance, wouldn't it? Or do I miss something?

Reply:
Essentially yes, although making them package private would be sufficient.

Personally I'd use a different route, though: give the factory an initialize(Context) method which creates the factory and assigns an `sInstance` variable; `getInstance` then simply returns that variable.